src/docker/composeappengine.cc

```diff
@@ -193,4 +315,78 @@ void ComposeAppEngine::extractAppArchive(const App& app, const std::string& arch
   }
 }
 
+ComposeAppEngine::AppState::AppState(const App& app, const boost::filesystem::path& root, bool set_version) try
+    : version_file_ {
+  (root / MetaDir / VersionFile).string()
+}
+, state_file_{(root / MetaDir / StateFile).string()} {
+  version_ = version_file_.readStr();
+  if (version_ == app.uri) {
+    state_ = static_cast<State>(state_file_.read());
+    return;
+  }
+
+  state_ = AppState::State::kUnknown;
+  if (set_version) {
+    version_ = app.uri;
+    version_file_.write(version_);
+  }
+}
+catch (const std::exception& exc) {
+  LOG_ERROR << "Failed to read version or state file: " << exc.what();
+}
+
+void ComposeAppEngine::AppState::setState(const State& state) {
+  state_file_.write(static_cast<int>(state));
+  state_ = state;
+}
+
+void ComposeAppEngine::AppState::File::write(int val) { write(&val, sizeof(val)); }
+
+void ComposeAppEngine::AppState::File::write(const std::string& val) { write(val.data(), val.size()); }
+
+int ComposeAppEngine::AppState::File::read() const {
+  int val{0};
+  read(&val, sizeof(val));
+  return val;
+}
+
+std::string ComposeAppEngine::AppState::File::readStr() const {
+  std::array<char, 4096> buf{0};
+  read(&buf, buf.size());
+  return buf.data();
+}
+
+int ComposeAppEngine::AppState::File::open(const char* file) {
+  int fd = ::open(file, O_CREAT | O_RDWR | O_SYNC, S_IRUSR | S_IWUSR);
+  if (-1 == fd) {
+    throw std::system_error(errno, std::system_category(), std::string("Failed to open/create a file: ") + file);
+  }
+  return fd;
+}
+
+void ComposeAppEngine::AppState::File::write(const void* data, ssize_t size) {
+  const std::string tmp_file{path_ + ".tmp"};
+  int fd = open(tmp_file.c_str());
+  int wr = ::write(fd, data, size);
+  if (-1 == wr) {
+    close(fd);
+    throw std::system_error(errno, std::system_category(), std::string("Failed to read from a file: ") + path_);
+  }
+  fsync(fd);
+  close(fd);
+
+  ::rename(tmp_file.c_str(), path_.c_str());
+  ::remove(tmp_file.c_str());
```

Comment:
remove shouldn't be needed unless rename failed and we ignore that, so i'm no sure its worth calling remove?

Reply:
Yes, I actually forgot to check the return value for `rename`.